asteroid_filterbanks/pcen.py

```diff
@@ -0,0 +1,162 @@
+from torch import nn
+import torch
+from . import transforms
+from typing import Union
+
+try:
+    from typing import TypedDict
+except ImportError:  # Fallback for >= Python 3.7
+    from typing_extensions import TypedDict
+
+
+class ExponentialMovingAverage(nn.Module):
+    """
+    Computes the exponential moving average of an sequential input.
+
+    Influenced by leaf-audio's tensorflow implementation
+        https://github.com/google-research/leaf-audio/blob/7ead2f9fe65da14c693c566fe8259ccaaf14129d/leaf_audio/postprocessing.py#L27
+
+    See license here
+        https://github.com/google-research/leaf-audio/blob/master/LICENSE
+    """
+
+    def __init__(
+        self,
+        smooth: float = 0.04,
+        per_channel: bool = False,
+        n_channels: int = 2,
+        trainable: bool = False,
+    ):
+        super().__init__()
+        if per_channel:
+            self.weights = nn.Parameter(
+                torch.full((n_channels,), fill_value=smooth), requires_grad=trainable
+            )
+        else:
+            self.weights = nn.Parameter(
+                torch.full((1,), fill_value=smooth), requires_grad=trainable
+            )
+
+    def forward(self, mag_spec, initial_state):
+        weights = torch.clamp(self.weights, 0.0, 1.0)
+        accumulator = initial_state
+        out = None
+        for x in torch.split(mag_spec, 1, dim=-1):
+            accumulator = weights * x + (1.0 - weights) * accumulator
+            if out is None:
+                out = accumulator
+            else:
+                out = torch.cat((out, accumulator), dim=-1)
+        return out
+
+
+class TrainableParameters(TypedDict):
+    alpha: bool
+    delta: bool
+    root: bool
+    smooth: bool
+
+
+class PCEN(nn.Module):
+    """Per-Channel Energy Normalization as described in [1].
+
+    This applies a fixed or learnable normalization by an exponential moving average smoother, and a compression.
+
+    Args:
+        alpha: The AGC strength (or gain normalization strength) (α in the paper).
+            Defaults to 0.96
+        delta: Bias added before compression (δ in the paper).
+            Defaults to 2.0
+        root: One over exponent applied for compression (r in the paper).
+            Defaults to 2.0
+        floor: Offset added to compression, prevents division by zero. (ϵ in the paper)
+            Defaults to 1e-6
+        smooth: Smoothing coefficient (s in the paper).
+            Defaults to 0.04
+        n_channels: Number of channels in the time frequency representation.
```

Comment:
IIUC, this is only used if `per_channel_smoothing` is used, right? 
If yes, add a note here please.

Reply:
It is not the case. It only applies to the smoothing (EMA) but not the other parameters (i.e. root, delta, alpha).

Reply:
Ok my bad.